google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+      : retry_policy_prototype_(std::move(retry_policy)),
+        backoff_policy_prototype_(std::move(backoff_policy)),
+        sleeper_(std::move(sleeper)),
+        stream_factory_(std::move(stream_factory)),
+        reinitializer_(std::move(reinitializer)),
+        impl_(stream_factory_()) {}
+
+  ResumableAsyncStreamingReadWriteRpcImpl(ResumableStreamingReadRpc&&) = delete;
+  ResumableAsyncStreamingReadWriteRpcImpl& operator=(
+      ResumableStreamingReadRpc&&) = delete;
+
+  void Cancel() override {
+    // not guarding since this is the singular writer
+    is_canceled_ = true;
+    std::lock_guard<std::mutex> g{mu_};
+    impl_->Cancel();
+  }
+
+  future<bool> Start() override {
+    std::lock_guard<std::mutex> g{mu_};
+    return impl_->Start().then([this](future<bool> start_fu) {
+      if (!start_fu.get()) {
+        std::shared_ptr<RetryPolicy const> retry_policy =
+            this->retry_policy_prototype_->clone();
+        std::shared_ptr<BackoffPolicy const> backoff_policy backoff_policy =
+            this->backoff_policy_prototype_->clone();
+        return this->RetryLoop(retry_policy, backoff_policy, RetryType.Start);
+      }
+      return make_ready_future(true);
+    });
+  }
+
+  future<absl::optional<Response>> Read() override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (!read_future_.has_value()) {
+      ProcessRead();
+    }
+    return read_future_.value();
+  }
+
+  future<bool> Write(Request const& r, grpc::WriteOptions o) override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (!write_future_.has_value()) {
+      ProcessWrite(r, o);
+    }
+    return write_future_.value();
+  }
+
+  future<bool> WritesDone() override {
+    std::lock_guard<std::mutex> g{mu_};
+    return impl_->WritesDone().then([this](future<bool> writes_done_fu) {
+      bool writes_done = writes_done_fu.get();
+      std::lock_guard<std::mutex> g{this->mu_};
+      impl_->Finish().then([this](future<Status> finish_future) {
+        Status finish_status = finish_future.get();
+        if (finish_status.ok()) {
+          this->finish_status_ = absl::make_optional(finish_status);
+          return make_ready_future(true);
+        }
+        auto const retry_policy = this->retry_policy_prototype_->clone();
+        auto const backoff_policy = this->backoff_policy_prototype_->clone();
+        return this->RetryLoop(retry_policy, backoff_policy, RetryType.Write);
+      });
+    });
+  }
+
+  future<Status> Finish() override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (finish_status_.has_value()) {
+      return make_ready_future(finish_status_.value());
+    }
+    return impl_->Finish();
+  }
+
+  future<Status> Status() {
+    std::lock_guard<std::mutex> g{mu_};
+    return make_ready_future(current_status_);
+  }
+
+ private:
+  enum class RetryType { Status, Read, Write };
+
+  absl::variant<future<absl::optional<ResponseType>>, future<bool>> RetryLoop(
+      std::shared_ptr<RetryPolicy const> retry_policy,
```

Comment:
these both need to be non-const.

In particular, I think there should be three functions for retries:

void OnFailure(Status status) {
  set up variables, call AttemptRetry
}

void AttemptRetry(Status status, std::shared_ptr<RetryPolicy> retry_policy, std::shared_ptr<BackoffPolicy> backoff_policy) {
   1) test if we should retry based on status
   2) sleep for sleep duration
   3) call Initialize
}

void Initialize(std::shared_ptr<RetryPolicy> retry_policy, std::shared_ptr<BackoffPolicy> backoff_policy) {
  ... Do logic for when we're allowed to retry ...
}

Then you can unify the start case as a call to Initialize